device/cuda/src/finding/finding_algorithm.cu

```diff
@@ -286,76 +252,111 @@ finding_algorithm<stepper_t, navigator_t>::operator()(
 
-        // Reset the number of tracks per seed
-        m_copy.memset(n_tracks_per_seed_buffer, 0)->ignore();
+    {
+        bound_track_parameters_collection_types::device in_params(
+            in_params_buffer);
+        bound_track_parameters_collection_types::const_device seeds(
+            seeds_buffer);
+        thrust::copy(thrust::cuda::par.on(stream), seeds.begin(), seeds.end(),
+                     in_params.begin());
+    }
 
-        // Global counter object: Device -> Host
-        TRACCC_CUDA_ERROR_CHECK(
-            cudaMemcpyAsync(&global_counter_host, global_counter_device.get(),
-                            sizeof(device::finding_global_counter),
-                            cudaMemcpyDeviceToHost, stream));
+    // Number of tracks per seed
+    vecmem::data::vector_buffer<unsigned int> n_tracks_per_seed_buffer(
+        n_seeds, m_mr.main);
 
-        m_stream.synchronize();
+    // Create a map for links
+    std::map<unsigned int, vecmem::data::vector_buffer<candidate_link>>
+        link_map;
 
-        // Set the number of input parameters
-        const unsigned int n_in_params = (step == 0)
-                                             ? in_params_buffer.size()
-                                             : global_counter_host.n_out_params;
+    // Create a map for parameter ID to link ID
+    std::map<unsigned int, vecmem::data::vector_buffer<unsigned int>>
+        param_to_link_map;
 
-        // Terminate if there is no parameter to process.
-        if (n_in_params == 0) {
-            break;
-        }
+    // Create a map for tip links
+    std::map<unsigned int, vecmem::data::vector_buffer<
+                               typename candidate_link::link_index_type>>
+        tips_map;
 
-        // Reset the global counter
-        TRACCC_CUDA_ERROR_CHECK(
-            cudaMemsetAsync(global_counter_device.get(), 0,
-                            sizeof(device::finding_global_counter), stream));
+    // Link size
+    std::vector<std::size_t> n_candidates_per_step;
+    n_candidates_per_step.reserve(m_cfg.max_track_candidates_per_track);
+
+    std::vector<std::size_t> n_parameters_per_step;
+    n_parameters_per_step.reserve(m_cfg.max_track_candidates_per_track);
+
+    unsigned int n_in_params = n_seeds;
+
+    for (unsigned int step = 0;
+         step < m_cfg.max_track_candidates_per_track && n_in_params > 0;
+         step++) {
 
         /*****************************************************************
          * Kernel2: Apply material interaction
          ****************************************************************/
 
-        nThreads = m_warp_size * 2;
-        nBlocks = (n_in_params + nThreads - 1) / nThreads;
-        kernels::apply_interaction<detector_type>
-            <<<nBlocks, nThreads, 0, stream>>>(det_view, m_cfg, n_in_params,
-                                               in_params_buffer);
-        TRACCC_CUDA_ERROR_CHECK(cudaGetLastError());
+        {
```

Comment:
I suppose theses parentheses are for readibility?

Reply:
Yes, and also for scopes, so we can redeclare `nThreads` and `nBlocks` for the next algorithms.

Reply:
OK this makes sense